### ISSUE TITLE
Assert failure in optimize.c line 7237

### ISSUE THREAD

**Issue by bennn** [racket/racket#1673]
Good afternoon,

This morning I made a small mistake, I wrote `(in-list 10)` basically. When I compiled my buggy program I got the error:

```
Assertion failed: (tvar->optimize_used), function flip_transitive, file ../../../racket/gc2/../src/optimize.c, line 7237.
```

Below is a smaller version of my program that makes the same error message. Run `$ raco make FILE.rkt` to test.

The program is very delicate. Inlining things, removing the `contract-out`, removing the call to `void` on the bottom line, e.g. all make the program compile.

```
#lang racket/base

(module benchmark-info racket/base
  (require racket/contract)
  (provide
    (contract-out
      [benchmark->num-configurations (-> any/c any)]
      [->benchmark-info (-> any/c any)]))

  (define (benchmark->num-configurations bm)
    1)

  (define (->benchmark-info bm-name)
    1)
)
(require 'benchmark-info)


;; =============================================================================

(define *sample-rate* (make-parameter 10))

(define (num-configurations->sample-size nc rate)
  (+ rate nc))

(define (make-lambda cfg)
  (lambda ()
    (map values cfg)))

(define (generate-samples bm-name)
  (define bm (->benchmark-info bm-name))
  (define random-configuration
    (make-lambda '(4)))
  (define sample-size
    (num-configurations->sample-size (benchmark->num-configurations bm) (*sample-rate*)))
  (for ([i (in-range 5)])
    (with-output-to-file "test"
      #:exists 'replace
      (lambda ()
        (for ([s (in-list sample-size)]) ;; MY MISTAKE: sample-size is a natural number, should be using `in-range`
          (void (random-configuration)))))))
```

**Comment by LeifAndersen** [comment]
FWIW, I've compressed the example a bit:

```
#lang racket/base

(module benchmark-info racket/base
  (require racket/contract)
  (provide (contract-out [->benchmark-info any/c]))

  (define ->benchmark-info #f))
(require 'benchmark-info)

(define *sample-rate* (make-parameter #f))

(define (num-configurations->sample-size nc rate)
  (+ nc))

(define (make-lambda cfg)
  (map values cfg))

(define (generate-samples bm-name)
  (void (->benchmark-info bm-name))
  (define random-configuration (make-lambda '()))
  (define sample-size
    (num-configurations->sample-size (*sample-rate*) (*sample-rate*)))
  (for ([i 0])
    (for ([s (in-list sample-size)])
      (void (random-configuration)))))
```

**Comment by LeifAndersen** [comment]
FWIW, I made it smaller still:

```
#lang racket/base

(module benchmark-info racket/base
  (require racket/contract)
  (provide (contract-out [->benchmark-info any/c]))
  (define ->benchmark-info #f))
(require 'benchmark-info)

(define *sample-rate* (make-parameter))
(define (num-configurations->sample-size)
  (void *sample-rate*))
(define (make-lambda cfg)
  (map values cfg))
(define (generate-samples)
  (define random-configuration (make-lambda #f))
  (define sample-size
    (num-configurations->sample-size))
  (for ([i 0])
    (for ([s (in-list sample-size)])
      (->benchmark-info random-configuration))))
```

**Comment by LeifAndersen** [comment]
And...one more time...

```
#lang racket/base

(module benchmark-info racket/base
  (require racket/contract)
  (provide (contract-out [->benchmark-info any/c]))
  (define ->benchmark-info #f))
(require 'benchmark-info)

(define *sample-rate* (void))
(define (num-configurations->sample-size)
  *sample-rate*)
(define make-lambda map)
(define (generate-samples)
  (define random-configuration (make-lambda ))
  (define sample-size
    (num-configurations->sample-size))
  (for ([i 0])
    (for ([s (in-list sample-size)])
      (->benchmark-info random-configuration))))
```

**Comment by mflatt** [comment]
Thanks! At this point, I'm pretty sure the repair is to simply remove an incorrect assertion, but I'll put a little more effort into getting a direct test case – so this smaller example is a helpful starting point.

**Comment by LeifAndersen** [comment]
Okay then. In that case I have one more (slightly) smaller example if that helps.

```
#lang racket/base

(module benchmark-info racket/base
  (require racket/contract)
  (provide (contract-out [+ any/c])))
(require 'benchmark-info)

(define *sample-rate* (void))
(define (num-configurations->sample-size)
  *sample-rate*)
(define (generate-samples)
  (define random-configuration (map))
  (define sample-size
    (num-configurations->sample-size))
  (for ([i 0])
    (for ([s (in-list sample-size)])
      (+ random-configuration))))
```

**Comment by mflatt** [comment]
Thanks, but I'm down to
```
#lang racket/base

(define unknown #f)
(set! unknown unknown)

(define (generate-samples)
  (define random-configuration (unknown))
  (for ([i 0])
    (for ([s (in-list 'obviously-not-a-list)])
      (unknown random-configuration))))
```

**Comment by mflatt** [comment]
Fixed in a71d87c5a9913cd20399e92536959d5b4de3fb3d
